### DISCUSSION TITLE
How to fetch user's JWT token from backend asynchronously

### DISCUSSION THREAD

**Discussion by amit3vedi** [Iterable/iterable-android-sdk#724] [Q&A]
Hello. i would like to know if there is any way to fetch JWT token asynchronously? as per the documentation i'm using following code to setup `IterableAuthHandler` 
```
.setAuthHandler(
                object : IterableAuthHandler {
                    override fun onAuthTokenRequested(): String {
                        GlobalScope.launch {
                            return AuthManager.getIdTokenAsync().await()
                        }
                      return ""
                    }

                    override fun onTokenRegistrationSuccessful(p0: String?) {
                    }

                    override fun onTokenRegistrationFailed(p0: Throwable?) {
                    }
                }
            )
```
Here `AuthManager.getIdTokenAsync()` is an API call which will take few seconds to return response. but this callback function need string as a return type. so how is it possible to return token after i receive it from backend ?

**Comment by Ayyanchira** [comment]
For async auth handling, we recommend using helper method - [`setAuthToken`](https://github.com/Iterable/iterable-android-sdk/blob/f925e44d169142d3e7e235b2edff31b4948cea98/iterableapi/src/main/java/com/iterable/iterableapi/IterableApi.java#L768-L770) whenever app fetches the authToken. 
However, it is to be noted that empty or null tokens will be passed in API calls that SDK make which will fail and appropriate failure handler will be called unless a valid auth token is passed.


In its current architecture, the SDK does not wait for async operation. 
However, `onAuthTokenRequested()` although being a blocking call, runs on a separate thread. So UI will not be blocked. 
Some alternative approach to consider is: 
To have the token pre-fetched separately as soon as user logs in, so its ready to be provided when SDK needs one.

**Reply by amit3vedi** [reply]
@Ayyanchira is there any way app can ask iterable SDK to retry sync once token is refreshed? currently i'm noticing issue where if token is expired during sync SDK is not automatically retry syncing after new token is generated. is there any way to manually initiate sync?
